### PR TITLE
S3 bucket policies and versioning

### PR DESCRIPTION
**Pull request by agalamatis**
trello: https://trello.com/c/w4tnuB26/505-enable-s3-bucket-versioning
also: https://trello.com/c/kyFOBl0M/512-allow-only-encrypted-connection-to-s3-buckets

### CODE REVIEW

File: terraform/environments/preview/s3_buckets.tf [hunk 1]
```diff
@@ -306,9 +381,57 @@ resource "aws_s3_bucket" "g7-draft-documents_bucket" {
   }
 }
 
+resource "aws_s3_bucket_policy" "g7-draft-documents_bucket" {
+  bucket = "${aws_s3_bucket.g7-draft-documents_bucket.id}"
+
+  policy = <<POLICY
```
Comment (katstevens) [inline terraform/environments/preview/s3_buckets.tf]: Is there a reason we're using an inline policy here over a `policy_document`? I think we want to use the latter where possible.

Comment (agalamatis) [reply terraform/environments/preview/s3_buckets.tf]: I was following the convention of existing policies, if they were inlined I left them inlined. But I can change them, if everyone prefers the TF syntax over inlined JSON.

**Comment by katstevens** [comment]
To check my understanding - are these `Deny` policies intended to disable everything on the baseline set of permissions, and we then add specific `Allow` policies on top of that? How do we manage buckets that only have a `Deny` policy?

(I'd definitely squash the terraformat linting commits, by the way.)

**Comment by agalamatis** [comment]
> To check my understanding - are these `Deny` policies intended to disable everything on the baseline set of permissions, and we then add specific `Allow` policies on top of that? How do we manage buckets that only have a `Deny` policy?

Yes, from what I understand all statements work in conjunction. Normally list bucket is denied to all, therefore we need to allow it to certain users. Non-encrypted connections are allowed, therefore we need to deny it.

> (I'd definitely squash the terraformat linting commits, by the way.)

I will

**Comment by katstevens** [comment]
@agalamatis OK, that makes sense. What happens when we add the `Deny` policy to say, the cross region replication buckets, which currently don't have any `Allow` policies set?

**Comment by agalamatis** [comment]
> @agalamatis OK, that makes sense. What happens when we add the `Deny` policy to say, the cross region replication buckets, which currently don't have any `Allow` policies set?

@katstevens The statements of a policy change the default behaviour of buckets. A new bucket accepts both HTTP and HTTPS, so I added a statement that denies non-encrypted connections. You think that this will have some kind of side-effect?

**Pull request merged**